lib/haskell/natural4/src/LS/XPile/MathLang/GenericMathLang/TranslateL4.hs

```diff
@@ -787,100 +813,118 @@ baseExpifyMTEs (splitGenitives -> (Just g, rest@(_:_))) = do
+                (False, False) -> throwNotSupportedWithMsgError (RPMT mtes) "baseExpifyMTEs: trying to apply non-function"
+
+        _mtes ->
+          case break (isFun userFuns) mtes of
+          -- function, rest
+            ([],f:ys) -> do
+              arg <- expifyMTEsNoMd ys
               assumedVar <- varFromMTEs [f]
               fExp <- mkVarExp assumedVar
-              let fArg1 = noExtraMdata $ EApp fExp arg1
-              pure $ EApp fArg1 arg2
-
-      -- no userfuns here
-        (_xs,[]) -> do
-          let parenExp = MTT $ textifyMTEs $ parenExps mtes
-          expParsedAsText <- parseExpr parenExp
-          case expParsedAsText of
-            ELit _ -> trace [i|parseExpr returned this as a string literal: #{expParsedAsText}|] do
-            -- TODO: this should definitely not be a sequence, what should it be instead???
-              parsedExs <- traverse parseExpr mtes
-              pure $ ESeq $ foldr consSeqExp mempty parsedExs
-
-            -- arithmetic expression like [MTT "m1",MTT "*",MTT "m2"]
-            notStringLit -> pure notStringLit
+              pure $ EApp fExp arg
+
+          -- mte1 [, …, mteN], function, rest
+            (xs,f:ys) -> do
+              arg1 <- expifyMTEsNoMd xs
+              arg2 <- expifyMTEsNoMd ys
+              -- since f is in userfuns, we can parse it using parseExpr. replacing args to some dummy x and y (but will fail if there is a userfun called x or y ¯\_(ツ)_/¯)
+              bexp <- parseExpr $ MTT [i|x #{mtexpr2text f} y|]
+              case bexp of
+                ENumOp {} -> pure $ ENumOp bexp.numOp arg1 arg2
+                ECompOp {} -> pure $ ECompOp bexp.compOp arg1 arg2
+                _ -> do
+                  assumedVar <- varFromMTEs [f]
+                  fExp <- mkVarExp assumedVar
+                  let fArg1 = noExtraMetadata $ EApp fExp arg1
+                  pure $ EApp fArg1 arg2
+
+          -- no userfuns here
+            (_xs,[]) -> do
+              let parenExp = MTT $ textifyMTEs $ parenExps mtes
+              expParsedAsText <- parseExpr parenExp
+              case expParsedAsText of
+                ELit _ -> trace [i|parseExpr returned this as a string literal: #{expParsedAsText}|] do
+                -- TODO: this should definitely not be a sequence, what should it be instead???
+                  parsedExs <- traverse parseExpr mtes
+                  pure $ ESeq $ foldr consSeqExp mempty parsedExs
+
+                -- arithmetic expression like [MTT "m1",MTT "*",MTT "m2"]
+                notStringLit -> pure notStringLit
   where
     consSeqExp :: BaseExp -> SeqExp -> SeqExp
-    consSeqExp = consSE . noExtraMdata
+    consSeqExp = consSE . noExtraMetadata
 
     parenExps :: [MTExpr] -> [MTExpr]
     parenExps mtexpr
       | any (isOp . mtexpr2text) mtexpr = parenNestedExprs <$> mtexpr
       | otherwise = mtexpr
 
     isOp :: T.Text -> Bool
-    isOp = (`elem` arithOps)
+    isOp = (`Set.member` arithOps)
```

Comment:
I wonder if it might make sense to use optics here to 'program more to an interface' / avoid programming to the particular implementation of sets used here.

Reply:
I don't feel strongly about that :shrug:. If we don't permit multiple implementations, feels like premature abstraction.

Reply:
Yeah, I don't have strong feelings about this; was just a quick thought